lib/ecto/parameterized_type.ex

```diff
@@ -99,6 +99,9 @@ defmodule Ecto.ParameterizedType do
   """
   @callback init(opts :: opts()) :: params()
 
+  @callback change(old_value :: any(), new_value :: any(), params :: params()) ::
+              {:ok, new_value :: any()} | {:error, Ecto.Changeset.error()}
```

Comment:
Perhaps it is best to make the last argument be a keyword list where you extract the message, similar to how we handle errors from types on `cast` here:

```elixir
      {:invalid, custom_errors} ->
        {message, new_errors} =
          custom_errors
          |> Keyword.put_new(:validation, :cast)
          |> Keyword.put(:type, type)
          |> Keyword.pop(:message, "is invalid")
        {changes, [{key, {message, new_errors}} | e
```

Reply:
@josevalim I understand that you are talking about `{:error, Ecto.Changeset.error()}`, right? If so then please take a look at that type definition:

```elixir
@type error() :: {String.t(), Keyword.t()}
```
I used that `typespec`, because that's how the changesets works. Look at last line you quoted and how I'm handling errors:
> ```elixir
>       {:error, error} ->
>         {changes, [{key, error} | errors], false}
> ```
> Source: https://github.com/Eiji7/ecto/blob/4e0838d98c037ea49e468ae8c48c7999d2d3630d/lib/ecto/changeset.ex#L1204-L1205

Or maybe by last argument you mean `params :: params()`, so I should (for now) pass something like `[type: {:parameterized, type, params}]` instead of just `params` (of parameterized type) to the callback?

Reply:
> I used that typespec, because that's how the changesets works. Look at last line you quoted and how I'm handling errors:

I am aware but we already have a precedent for returning errors from `Ecto.Type` and they are not of the type `Ecto.Changeset.error`. Instead they are a keyword list. Please look at Ecto.Type.cast return type. :)

Reply:
done